pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSource.java

```diff
@@ -60,17 +74,118 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
+            for (String transformName : transformsList.split(",")) {
+                transformName = transformName.trim();
+                String prefix = "transforms." + transformName + ".";
+                String typeKey = prefix + "type";
+                Object classNameObj = config.get(typeKey);
+                if (classNameObj == null) {
+                    continue;
+                }
+                String className = classNameObj.toString();
+                try {
+                    @SuppressWarnings("unchecked")
+                    Class<Transformation<SourceRecord>> clazz =
+                        (Class<Transformation<SourceRecord>>) Class.forName(className);
+                    Transformation<SourceRecord> transform = clazz.getDeclaredConstructor().newInstance();
+                    java.util.Map<String, Object> transformConfig = config.entrySet().stream()
+                        .filter(e -> e.getKey().startsWith(prefix))
+                        .collect(java.util.stream.Collectors.toMap(
+                            e -> e.getKey().substring(prefix.length()),
+                            java.util.Map.Entry::getValue
+                        ));
+                    log.info("transform config: {}", transformConfig);
+                    String predicateName = (String) transformConfig.get("predicate");
+                    boolean negated = Boolean.parseBoolean(
+                        String.valueOf(transformConfig.getOrDefault("negate", "false")));
+                    Predicate<SourceRecord> predicate = null;
+                    if (predicateName != null) {
+                        predicate = predicates.get(predicateName);
+                        if (predicate == null) {
+                            log.warn("Transform {} references non-existent predicate: {}",
+                                    transformName, predicateName);
+                        }
+                    }
+                    transform.configure(transformConfig);
+                    transformations.add(new PredicatedTransform(predicate, transform, negated));
+                } catch (Exception e) {
+                    throw new RuntimeException("Failed to instantiate SMT: " + className, e);
+                }
+            }
+        }
+    }
+
+    private static final AvroData avroData = new AvroData(1000);
 
     public synchronized KafkaSourceRecord processSourceRecord(final SourceRecord srcRecord) {
-        KafkaSourceRecord record = new KafkaSourceRecord(srcRecord);
-        offsetWriter.offset(srcRecord.sourcePartition(), srcRecord.sourceOffset());
+        SourceRecord transformedRecord = applyTransforms(srcRecord);
+        if (transformedRecord == null) {
+            return null;
+        }
+        KafkaSourceRecord record = new KafkaSourceRecord(transformedRecord);
+        offsetWriter.offset(transformedRecord.sourcePartition(), transformedRecord.sourceOffset());
```

Comment:
would it better to keep this as `offsetWriter.offset(srcRecord.sourcePartition(), srcRecord.sourceOffset())` ?

Reply:
Makes sense. Otherwise, it may result in unnecessary re-deliveries of messages that will anyways not be ingested. I'll address this shortly.